File: frontend/src/models/Application.js
import mongoose from 'mongoose';

const TimelineEntry = new mongoose.Schema({
  date: { type: Date, default: Date.now },
  event: { type: String, required: true },
  notes: { type: String, default: '' },
}, { _id: true });

const ApplicationSchema = new mongoose.Schema({
  userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
  company: { type: String, required: true, trim: true },
  role: { type: String, required: true, trim: true },
  type: {
    type: String,
    enum: ['full-time', 'internship', 'contract', 'freelance'],
    default: 'full-time',
  },
  status: {
    type: String,
    enum: ['bookmarked', 'applied', 'screening', 'technical', 'onsite', 'offer', 'accepted', 'rejected', 'ghosted'],
    default: 'bookmarked',
  },
  appliedDate: { type: Date, default: null },
  salary: { type: String, default: '' },
  location: { type: String, default: '' },
  url: { type: String, default: '' },
  notes: { type: String, default: '', maxlength: 2000 },
  linkedFileId: { type: mongoose.Schema.Types.ObjectId, ref: 'StudyFile', default: null },
  linkedCaptureId: { type: mongoose.Schema.Types.ObjectId, ref: 'Capture', default: null },
  timeline: [TimelineEntry],
}, { timestamps: true });

ApplicationSchema.index({ userId: 1, status: 1 });
ApplicationSchema.index({ userId: 1, company: 1 });

export default mongoose.models.Application || mongoose.model('Application', ApplicationSchema);

File: backend-spring/src/main/java/com/nextlevel/api/model/Application.java
package com.nextlevel.api.model;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.mapping.Document;

import lombok.Data;

@Data
@Document(collection = "applications")
@CompoundIndexes({
        @CompoundIndex(def = "{'userId': 1, 'status': 1}"),
        @CompoundIndex(def = "{'userId': 1, 'company': 1}")
})
public class Application {

    @Data
    public static class TimelineEntry {
        private Instant date = Instant.now();
        private String event;
        private String notes = "";
    }

    @Id
    private String id;
    private String userId;
    private String company;
    private String role;
    private String type = "full-time";
    private String status = "bookmarked";
    private Instant appliedDate;
    private String salary = "";
    private String location = "";
    private String url = "";
    private String notes = "";
    private String linkedFileId;
    private String linkedCaptureId;
    private List<TimelineEntry> timeline = new ArrayList<>();
    private Instant createdAt;
    private Instant updatedAt;
}
